From 63bd2a12c30109cb4a3501e504a8b7b26d1749d5 Mon Sep 17 00:00:00 2001
From: Arnon Kaufman <akaufman@habana.ai>
Date: Sun, 21 Feb 2021 09:30:08 +0200
Subject: [PATCH] [SW-32519] i2c_slave: fix module disable

The i2c disable procedure was not working properly.
And the i2c bus stopped responding. (i2cdetect could not see info)

By adding delay and toggling the psoc i2c enable. as the i2c teardown
function is doing. The issue was fixed.

Change-Id: Id833666fa3755e8f279f128520dc191aa5f911d5
Signed-off-by: Arnon Kaufman <akaufman@habana.ai>
---

diff --git a/common/module_i2c_slave.c b/common/module_i2c_slave.c
index 9005e4b..3d093d8 100644
--- a/common/module_i2c_slave.c
+++ b/common/module_i2c_slave.c
@@ -198,10 +198,20 @@
 
 void module_i2c_s_disable(void)
 {
+	u32 val;
+
 	i2cslave_flush();
 	regmap_writel(0x0, mmPSOC_I2C_S_IC_ENABLE);
-	i2cslave_flush();
+	udelay(100);
 
+	/* Write IC_SLAVE_DISABLE to IC_CON */
+	val = regmap_readl(mmPSOC_I2C_S_IC_CON);
+	val |= PSOC_I2C_S_IC_CON_IC_SLAVE_DISABLE_MASK;
+	regmap_writel(val, mmPSOC_I2C_S_IC_CON);
+
+	udelay(1000);
+
+	i2cslave_flush();
 	preboot_puts_dbg("I2C SLAVE: disabled\n");
 }
 
